### PR TITLE
End of month events (#539)

### PR DESCRIPTION
**Pull request by pledbrook**
See second commit for details of the events added.

### CODE REVIEW

**Comment by Xymanek** [comment]
What's the point of `Monthly Report Closed`? Why not just use a UISL?

**Comment by pledbrook** [comment]
I don't know what the original reason was. I pretty much just ported as is in terms of the number of events. One thing this approach has in its favour is consistency in how and where the listeners are implemented. It makes it easy for mods to keep all their end-of-month processing together.

**Comment by pledbrook** [comment]
And if the event were to stay, it should probably be renamed to `PostEndOfMonth` for consistency with the first one.

**Comment by pledbrook** [comment]
I've justified to myself that the 'PostEndOfMonth' event is worth it because:

 * It balances nicely with 'PreEndOfMonth'
 * It's association with the monthly report screen is coincidental
 * It's better for mod authors if they have all the events they need provided in a consistent way - telling them to use a screen listener for this one aspect seems off
 * It allows mod authors to put all their end-of-month processing code in one place

**Comment by pledbrook** [comment]
@Musashi1584 I don't know if this has the green light to go. @Xymanek Any more feedback?

**Comment by Xymanek** [comment]
I kinda fail to understand what `NegativeMonthlyIncome` does. Perhaps a better name is in order?

**Comment by pledbrook** [comment]
I'll take another look at NegativeMonthlyIncome. I think it's combining two jobs into one, which makes it a bit confusing.

**Comment by pledbrook** [comment]
@Musashi1584 Did this get merged? If so, my mistake. Should have marked it as WIP.

**Comment by pledbrook** [comment]
@Xymanek I've split the 'NegativeMonthlyIncome' event into two. Hopefully the purpose of each of them is much clearer now.

**Comment by pledbrook** [comment]
I made a boo boo, which meant the 'OverrideSupplyDrop' event was never triggered. Merged the change into the previous commit and force pushed, since I don't want the other two commits squashed during the merge.

**Comment by pledbrook** [comment]
@Xymanek Was the change sufficient?

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersResistance.uc [hunk 1]
```diff
@@ -397,32 +403,83 @@ function GiveSuppliesReward(XComGameState NewGameState)
...
+	Tuple.Data[0].i = SupplyValue;  // The amount of negative supplies
+
+	`XEVENTMGR.TriggerEvent('ProcessNegativeIncome', Tuple, self, NewGameState);
 }
+// End Issue #539
 
 //---------------------------------------------------------------------------------------
-function int GetSuppliesReward(optional bool bUseSavedPercentDecrease)
+// Issue #539: Added the optional NewGameState argument
+function int GetSuppliesReward(optional bool bUseSavedPercentDecrease, optional XComGameState NewGameState)
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersResistance.uc]: This breaks BC policy.

If any mod MCOs `XComGameState_HeadquartersResistance` and implements this function, this change will cause a hard crash

Comment (pledbrook) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersResistance.uc]: Good catch. I've restored backwards compatibility by introducing `GetSuppliesReward_CH()` as the new function that takes a game state argument.

**Pull request merged**